x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotAllocator.java

```diff
@@ -207,6 +213,59 @@ public void allocateUnassigned(
         }
     }
 
+    /**
+     * @return the restore UUID to use when adjusting the recovery source of the shard to be a snapshot recovery source from a
+     * repository of the correct name. Returns {@code null} if the recovery source should not be adjusted.
+     */
+    @Nullable
+    private static String getRecoverySourceRestoreUuid(ShardRouting shardRouting, RoutingAllocation allocation) {
+        switch (shardRouting.recoverySource().getType()) {
+            case EXISTING_STORE:
+            case EMPTY_STORE:
+                // this shard previously failed and/or was force-allocated, so the recovery source must be changed to reflect that it will
+                // be recovered from the snapshot again
+                return RecoverySource.SnapshotRecoverySource.NO_API_RESTORE_UUID;
+            case SNAPSHOT:
+                // the recovery source may be correct, or we may be recovering it from a real snapshot (i.e. a backup)
+
+                final RecoverySource.SnapshotRecoverySource recoverySource = (RecoverySource.SnapshotRecoverySource) shardRouting
+                    .recoverySource();
+                if (recoverySource.restoreUUID().equals(RecoverySource.SnapshotRecoverySource.NO_API_RESTORE_UUID)) {
+                    // this shard already has the right recovery ID, but maybe the repository name is now different, so check if it needs
+                    // fixing up again
+                    return RecoverySource.SnapshotRecoverySource.NO_API_RESTORE_UUID;
+                }
+
+                // else we're recovering from a real snapshot, in which case we can only fix up the recovery source once the "real"
```

Comment:
I am not sure I understand this case, why can't we just always fix the repository name and leave the restore UUID alone when it is already a snapshot recovery source.

Reply:
Ah, that's because we trip over the `RestoreInProgressAllocationDecider` if the restore UUID in the recovery source doesn't match an ongoing restore. I pushed fa54a31b463 to explain this in the comment.

Reply:
Ahh, thanks for the explanation. I wish this was not necessary, but fixing that is likely a much bigger changes, so agree to this approach.